### PR TITLE
Adds fishing for real this time

### PR DESCRIPTION
**Pull request by ToasterBiome**
# Document the changes in your pull request

Inspired by https://terraria.fandom.com/wiki/Fishing

Adds fish, fishing rods, bait, the TackleBox 2000. The higher your fishing power (for right now, rod base power + whatever bait you have on it + CLOTHES BONUS) heighten your chances of getting a bite. Fish have weights and lengths, and you can show off your fish to other people.
(planned) You can get fish, little pieces of trash, small (SMALL) amounts of materials, whatever you fish up from the depths. Spawn a lone-op because captain was fishing in the same spot all shift. 
dreamseeker_JBAxDLNP76

you can butcher fish to be make fish fillets which are like the carpmeat but with no carpotoxin in it.

New vendor that sells fishing rods, bait, and new clothing that boosts your fishing power

dreamseeker_nLkvZbADRQ
dreamseeker_ehB9lT5Wl9
dreamseeker_t8KvhuskBf

https://user-images.githubusercontent.com/5091394/162081540-e7eb1a73-3014-4da6-9e60-99ada9c67227.mp4

new fish:

- galactic goldfish
- giant galactic goldfish
- space salmon
- big bang bass
- temporal tuna
- space shrimp
- space squid
- plasma pufferfish

You can cast from far away. If you're close, you will auto pick up the fish. Fishing loot is stored in a fishable component, which can be added to any turf. (Lava fishing anyone?) along with configurable loot tables.

Things left to do

- [X] reduce bobber sprite size
- [X] better bait sprites
- [X] bait overlay on fishing rods
- [X] cut up fish into fish meat
- [X] clothes that boost fishing power
- [ ] put in maps

For the future
- [ ] steal seafood recipes from other codebases
- [ ] more fishing rods
- [ ] expand fishable component to have different loot tables through fishing_power
- [ ] expand fishable component to work with other types
- [ ] expand water loot table

# Wiki Documentation

Fish, Rods, Bait, new vending machine, new mechanics.

# Changelog

:cl:  ToasterBiome, Tatax
rscadd: Adds fishing rods, fish, bait, new fishing vending machine
rscadd: Adds fishing
/:cl:

### CODE REVIEW

**Comment by Georgeur** [comment]
FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE FISH UPDATE

**Comment by Chubbygummibear** [comment]
IT'S TIME FOR FISH
pog fish

**Comment by Marmio64** [comment]
water literally does not exist as a tile outside of an icemoon ruin unless you count puddles and the swimming pool. You'll have to add it in somewhere and i'd suggest putting it in some ruins, especially ones that can be teleported to like the space bar, and winter biodome inn (this ones broken though im pretty sure). Can also add it to some other ruins for people to try to find if space bar does not spawn and perhaps a random maint variation. It'd be a little boring if it was just instantly accessible for everyone at any time such as if literally just a freshwater lake was shoved into every station.

**Comment by ilikesaying** [comment]
terraria fishing

**Comment by ToasterBiome** [comment]
> water literally does not exist as a tile outside of an icemoon ruin unless you count puddles and the swimming pool. You'll have to add it in somewhere and i'd suggest putting it in some ruins, especially ones that can be teleported to like the space bar, and winter biodome inn (this ones broken though im pretty sure). Can also add it to some other ruins for people to try to find if space bar does not spawn and perhaps a random maint variation. It'd be a little boring if it was just instantly accessible for everyone at any time such as if literally just a freshwater lake was shoved into every station.

As I said in the PR body, fishable is a component that can be added to any turf yogs wants.

**Comment by ghost** [comment]
Fish

**Review by ghost — APPROVED** [review]

**Comment by Moja10** [comment]
Badass, I hope some pond ruins, both in space and lavaland, can get added with these changes.

**Comment by cowbot92** [comment]
God I Just Want To Fish

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	if(!fc)
+		return ..()
+	if(!fishing)
+		if(!wielded)
+			to_chat(user,"You need to wield the rod in both hands before you can cast it!")
+			return
+		cast(fc,user)
+	else
+		if(fc != fishing_component)
+			to_chat(user,"You are not fishing here!")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
			to_chat(user, span_warning("You are not fishing here!"))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	. = ..()
+	. += "It's current fishing power is [fishing_power]."
+
+/obj/item/twohanded/fishingrod/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	var/datum/component/fishable/fc = target.GetComponent(/datum/component/fishable)
+	if(!fc)
+		return ..()
+	if(!fishing)
+		if(!wielded)
+			to_chat(user,"You need to wield the rod in both hands before you can cast it!")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
			to_chat(user, span_warning("You need to wield the rod in both hands before you can cast it!"))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		if(fc != fishing_component)
+			to_chat(user,"You are not fishing here!")
+			return
+		reel_in() //intentionally able to reel in with one hand
+
+/obj/item/twohanded/fishingrod/process()
+	if(!fishing)
+		return PROCESS_KILL
+	if(bite)
+		to_chat(fisher,"Whatever was on the line drifts back into the deep...")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
		to_chat(fisher, span_warning("Whatever was on the line drifts back into the deep..."))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+		bite = FALSE
+		return
+
+	if(prob(50))
+		to_chat(fisher,"Something bites! Reel it in!")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
		to_chat(fisher, span_boldnotice("Something bites! Reel it in!"))
```

could be normal notice

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	to_chat(fisher,"You cast out your fishing rod.")
+
+/obj/item/twohanded/fishingrod/proc/reel_in(var/forced = FALSE)
+	if(!forced && bite) // we got something!!!
+		playsound(fishing_component, 'sound/effects/water_emerge.ogg', 50, FALSE, -5)
+		spawn_reward()
+	else
+		playsound(fishing_component, 'sound/effects/splash.ogg', 50, FALSE, -5)
+		if(forced)
+			to_chat(fisher,"The auto-reel on the fishing rod activates!")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
			to_chat(fisher, span_boldnotice("The auto-reel on the fishing rod activates!"))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	fishing_component = fc
+	var/turf/fishing_turf = fishing_component.parent
+	fishing_turf.add_overlay(bobber)
+	fisher = user
+	RegisterSignal(user,COMSIG_MOVABLE_MOVED, .proc/reel_in_forced,TRUE)
+	RegisterSignal(src,COMSIG_MOVABLE_MOVED, .proc/reel_in_forced,TRUE)
+	RegisterSignal(src,COMSIG_ITEM_DROPPED, .proc/reel_in_forced,TRUE)
+	START_PROCESSING(SSobj,src)
+	playsound(fishing_component, 'sound/effects/splash.ogg', 50, FALSE, -5)
+	to_chat(fisher,"You cast out your fishing rod.")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
	to_chat(fisher, span_italics("You cast out your fishing rod..."))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+/obj/item/twohanded/fishingrod/proc/spawn_reward()
+	var/obj/picked_reward = fishing_component.getCommonLoot()
+	var/obj/reward_item = new picked_reward(fishing_component.parent)
+	reward_item.alpha = 0
+	reward_item.pixel_y = -12
+	animate(reward_item,time = 0.25 SECONDS,pixel_y = 0,alpha = 255,easing = SINE_EASING)
+	if(!fisher) //uh oh
+		return
+	visible_message("[fisher] reels in a [reward_item]!","You reel in a [reward_item]!")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
	fisher.visible_message(span_notice("[fisher] reels in a [reward_item]!"), span_notice("You reel in a [reward_item]!"))
```

could do the other span notice thingy

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+/obj/item/twohanded/fishingrod/proc/reel_in(var/forced = FALSE)
+	if(!forced && bite) // we got something!!!
+		playsound(fishing_component, 'sound/effects/water_emerge.ogg', 50, FALSE, -5)
+		spawn_reward()
+	else
+		playsound(fishing_component, 'sound/effects/splash.ogg', 50, FALSE, -5)
+		if(forced)
+			to_chat(fisher,"The auto-reel on the fishing rod activates!")
+		else
+			to_chat(fisher,"You reel in your fishing rod.")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
			to_chat(fisher, span_italics("You reel in your fishing rod."))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	if(!istype(B,/obj/item/reagent_containers/food/snacks/bait))
+		return
+
+	if(bait)
+		user.put_in_hands(bait)
+
+	if(!user.transferItemToLoc(B,src))
+		return
+	bait = B
+	to_chat(user, "You attach the [bait] to the fishing rod.")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
	to_chat(user, span_notice("You attach the [bait] to the fishing rod."))
```

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	cut_overlays()
+	add_overlay(bait)
+	recalculate_power()
+
+/obj/item/twohanded/fishingrod/AltClick(mob/living/user)
+	if(bait)
+		user.put_in_hands(bait)
+		to_chat(user, "You take the [bait] off the fishing rod.")
```
Comment (tattax) [inline yogstation/code/game/objects/items/fishing/rods.dm]:
```suggestion
		to_chat(user, span_notice("You take the [bait] off the fishing rod."))
```

Comment (tattax) [reply yogstation/code/game/objects/items/fishing/rods.dm]: god i love my sweet colored texts

File: yogstation/code/game/objects/items/fishing/rods.dm (hunk 6 above)
Comment (ToasterBiome) [reply yogstation/code/game/objects/items/fishing/rods.dm]: Find a span that does what you want, not using one that _almost_ does what you want and adding to it

File: yogstation/code/game/objects/items/fishing/rods.dm (hunk 8 above)
Comment (ToasterBiome) [reply yogstation/code/game/objects/items/fishing/rods.dm]: same here, NO!

File: yogstation/code/game/objects/items/fishing/rods.dm (hunk 9 above)
Comment (ToasterBiome) [reply yogstation/code/game/objects/items/fishing/rods.dm]: NOOOOOOOO

File: yogstation/code/game/objects/items/fishing/rods.dm (hunk 6 above)
Comment (tattax) [reply yogstation/code/game/objects/items/fishing/rods.dm]: `span_smallnoticeital("<FONT size = 2>You cast out your fishing rod..."))`

Comment (ToasterBiome) [reply yogstation/code/game/objects/items/fishing/rods.dm]: then edit your suggestion

Comment (tattax) [reply yogstation/code/game/objects/items/fishing/rods.dm]: this is better

Comment (tattax) [reply yogstation/code/game/objects/items/fishing/rods.dm]: do you like smallnoticeital or notice(`<i>`

File: code/modules/vending/fishing.dm [hunk 11]
```diff
@@ -0,0 +1,17 @@
+/obj/machinery/vending/fishing
+	name = "\improper Tackle Box 2000"	
+	desc = "Fishing peaked in 2000!"
+	icon = 'yogstation/icons/obj/vending.dmi'
+	icon_state = "fishing"
+	icon_deny = "fishing-deny"
+	product_slogans = "Don't tell my wife which bank I went to!;Fish fear me. Women love me.;If you want me to listen to you... talk about FISHING!;Good things come to those who bait.;Why did the lizard cross the road?"
+	vend_reply = "Go get em, tiger!"
```
Comment (Anvilman6) [inline code/modules/vending/fishing.dm]: "Go get 'em, tiger!"

File: yogstation/code/game/objects/items/fishing/fish.dm [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+/obj/item/reagent_containers/food/snacks/fish/proc/stop_overlay()
+	if(showoffer && showoff_overlay)
+		UnregisterSignal(showoffer,COMSIG_ATOM_DIR_CHANGE)
+		showoffer.cut_overlay(showoff_overlay)
+		showoffer = null
+		showoff_overlay = null
+
+/obj/item/reagent_containers/food/snacks/fish/goldfish
+	name = "galactic goldfish"
+	desc = "it's so... small!"
```
Comment (Anvilman6) [inline yogstation/code/game/objects/items/fishing/fish.dm:60]: Capitalization on fish descriptions? Fishing rod descs are capitalized, but I may've missed something on the fish setup.

File: yogstation/code/game/objects/items/fishing/fish.dm [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	desc = "you can tune a piano but you can't tuna fish"
+	icon_state = "fish_tuna"
+	min_length = 15
+	max_length = 79
+	min_weight = 18
+	max_weight = 1000
+
+/obj/item/reagent_containers/food/snacks/fish/shrimp
+	name = "space shrimp"
+	desc = "he looks a little shrimpy"
```
Comment (Anvilman6) [inline yogstation/code/game/objects/items/fishing/fish.dm:110]: Suggestion: "A shrimp! It's as shrimple as that."

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 14]
```diff
@@ -0,0 +1,149 @@
+/obj/item/twohanded/fishingrod
+	name = "fishing rod"
+	desc = "A fishing rod used for fishing. Despite looking ordinary, fishing has evolved for the cosmos with features like auto-reeling."
```
Comment (Anvilman6) [inline yogstation/code/game/objects/items/fishing/rods.dm]: Suggestion: "A rod used for fishing. Despite ordinary appearances, fishing has evolved to suit the cosmos with various features, like auto-reeling."

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+	var/static/mutable_appearance/bobber = mutable_appearance('yogstation/icons/obj/fishing/fishing.dmi',"bobber")
+	var/datum/component/fishable/fishing_component
+	var/mob/fisher
+	var/bite = FALSE
+
+	
+
+/obj/item/twohanded/fishingrod/examine(mob/user)
+	. = ..()
+	. += "It's current fishing power is [fishing_power]."
```
Comment (Anvilman6) [inline yogstation/code/game/objects/items/fishing/rods.dm]: "Its" - possessive, not a contraction

File: yogstation/code/game/objects/items/fishing/rods.dm [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+/obj/item/twohanded/fishingrod/proc/spawn_reward()
+	var/obj/picked_reward = fishing_component.getCommonLoot()
+	var/obj/reward_item = new picked_reward(fishing_component.parent)
+	reward_item.alpha = 0
+	reward_item.pixel_y = -12
+	animate(reward_item,time = 0.25 SECONDS,pixel_y = 0,alpha = 255,easing = SINE_EASING)
+	if(!fisher) //uh oh
+		return
+	fisher.visible_message(span_notice("[fisher] reels in a [reward_item]!"), span_notice("You reel in a [reward_item]!"))
```
Comment (Anvilman6) [inline yogstation/code/game/objects/items/fishing/rods.dm:123]: May need it to be a(n) depending on the names of rewards

**Review by Anvilman6 — COMMENTED** [review]
Grammar+Syntax nitpicking time because Biome is my favorite PR maker

**Comment by SomeguyManperson** [comment]
is this being worked on

**Comment by ToasterBiome** [comment]
> is this being worked on

Yes. Took a bit of a back seat when i was doing gax stuff.

**Comment by JamieD1** [comment]
@ToasterBiome This is maps now, I see its not ticked still. Just making sure.

**Comment by ToasterBiome** [comment]
Testmerging for a round teehee

**Review by JamieD1 — APPROVED** [review]

**Comment by ghost** [comment]
Sus

**Pull request merged**